### ISSUE TITLE
ToDo list

### ISSUE THREAD

**Issue by jarun** [jarun/nnn#935]
Rolled from #881.

### Cooking

- [x] <kbd>Bksp</kbd>/<kbd>Del</kbd> at empty filter prompt to refresh dir
- [x] fix zombies left behind after running plugins (#999)
- [x] consider nanosecond resolution when sorting by time (#978)
- [x] check external selection in `'c'urrent / 's'el` prompt (#976)
- [x] support picker plugins e.g. `fzopen` (https://github.com/mcchrish/nnn.vim/issues/82)
- [x] make batch rename interactive (#971)
- [x] cached uid/gid for performance improvement
- [x] support [QuickLook](https://github.com/QL-Win/QuickLook) on WSL in `preview-tui*` (#959)
- [x] toggle `preview-tui*` with the same _custom_ plugin key
- [x] smoother preview toggling in `preview-tui*` (#966)
- [x] config `listen_on` should be set in kitty.conf for `preview-tui*` (#970)
- [x] minimal `bat` style in `preview-tui*`, honors [BAT_STYLE](https://github.com/sharkdp/bat#output-style)
- [x] plugin `unmount-parent`: unmount a remote mountpoint from within
- [x] plugin `xdgdefault`: set the default application for hovered file type
- [x] plugin `fzplug`: fuzzy find, preview and run other plugins
- [x] plugin `cmusq`: queue/play music in `cmus`
- [x] plugin `mocplay` renamed to `mocq`
- [x] go to last dir on <kbd>~</kbd> (HOME) or <kbd>`</kbd> (ROOT) key repeat
- [x] key <kbd>^Space</kbd> for range selection/clear dropped, use <kbd>m</kbd> (#998)
- [x] a [Themes](https://github.com/jarun/nnn/wiki/Themes) wiki page for users to add and refer to
- [x] show selection mark (`+`) in reverse bold for improved visibility
- [x] reversed block instead of `>` to mark hovered entry in detail mode
- [x] make option `O_CKBOARD` retired in favour of reversed block cursor
- [x] retain filter in _nav-to-type_ mode after file open
- [x] script to statically compile with `netbsd-curses` + `musl` on Ubuntu
- [x] fixes for `netbsd-curses` + `musl` (https://github.com/sabotage-linux/netbsd-curses/issues/46)
- [x] fix broken picker mode with `NNN_TMPFILE` exported and <kbd>q</kbd> to quit
- [x] disable xterm title setting in picker mode (#974)
- [x]  make option `NOX11`: disable notis, sel->clipboard sync, xterm title

### Up for grabs

- [ ] multi-threaded disk usage calculation
- [ ] notcurses UI backend (#1014) [@KlzXS]
- [ ] a light colorscheme with `NNN_FCOLORS` (#937)
- [ ] update only current context data when saving session (#837)
- [ ] generate static binary for arm, aarch with releases using CI (#623)
- [ ] support predefined filters like bookmarks
- [ ] open new terminal window in spawned shell directory (#759)

For anything else please discuss in this thread.

**Contribution [guideline](https://github.com/jarun/nnn/blob/master/misc/CONTRIBUTING.md).**

**Comment by luukvbaal** [comment]
Following discussion from #267 I do think showing the the last characters of the file name should be prioritized in the case of file names longer than the name column. I was wondering if we should consider changing the `unescape()` function to return something like `abcmnopqrst.txt` or `abc...pqrst.txt` for `abcdefghijklmnopqrst.txt`.

**Comment by jarun** [comment]
Doesn't cover all cases inclusively. For some filenames the difference may be in the middle. There's always <kbd>f</kbd>/<kbd>^F</kbd> for file details. Just leave it be.

**Comment by lawnowner** [comment]
@jarun (https://github.com/jarun/nnn/issues/499#issuecomment-838266621) The prompt looks better than before with repl now. This morning, I got a three line error message starting with "`Assembler: `" after running `sleep 1; date` and typing stuff while asleep, although all I can do is wishing you good luck with reproducing that as I can't even remember the full error message, let alone the keys pressed. As somebody who has been using nnn in monochrome, and with almost none of the official plugins, I guess the next release will be my favourite, except for the tilde contraction next to the context numbers; it looks almost like nothing whem I'm home.

**Comment by jarun** [comment]
Get used to the tilde. It was requested by multiple users.

**Comment by Kabouik** [comment]
I noticed `preview-tui` and `preview-tui-ext` queue image previews when scrolling through multiple files. It is probably a non-issue on most machines, but breaks previews on slower ones (_i.e._, SBCs running on SD cards, machines/VM/LXC without hardware acceleration for kitty image preview, _etc._) because load times add up for every file hovered along the way.

It may be true for other file types, but is most visible for big images since they take longer to load (on my phone running Debian and previewing with kitty's `icat`, the issue is not huge with images around 500 KB, but very noticeable with images around 2-4 MB). I suppose even on real computers this could be an issue with large images from modern DSLRs or scanned files.

I tried: 

```patch
        kitty +kitten icat --silent --place "$1"x"$2"@0x0 --transfer-mode=stream --stdin=no \
-           "$3"
        kitty +kitten icat --silent --place "$1"x"$2"@0x0 --transfer-mode=stream --stdin=no \
+           "$3" &

```
Which seemed to limit queued previews, but resulted in previews being shown only if scrolling slowly.

**Should I open a dedicated issue about it?**

While I mention preview plugins, how about making default `bat` previews more minimal in with:

```patch
-           fifo_pager bat --terminal-width="$cols" --paging=never --decorations=always --color=always \
+           fifo_pager bat --terminal-width="$cols" --paging=never --style=numbers,header,plain --color=always \
```
It s a very minor difference and perhaps just personal taste, but I see no real advantage for the grid view by default:

![Screenshot_20210513_006](https://user-images.githubusercontent.com/7107523/118190659-bf496980-b443-11eb-9b7a-0c46bc32d677.png)
![Screenshot_20210513_005](https://user-images.githubusercontent.com/7107523/118190665-c1132d00-b443-11eb-807c-4dc1b1392e2b.png)

**Comment by jarun** [comment]
@leovilok @luukvbaal can you please take a look?

> It s a very minor difference and perhaps just personal taste, but I see no real advantage for the grid view by default

Yes, it actually looks awful to anyone from sane more/less background. Please raise a PR to remove this from all plugins.

**Comment by luukvbaal** [comment]
The preview plugins work with a FIFO so the previews are inherently queued. Not sure anything can be done about a slow machine that can't keep up with how fast you change the selection.

Never noticed the `bat` defaults since I have `BAT_STYLE` in my profile. I agree the grid is only clutter so I'm fine with getting rid of it.

**Comment by Kabouik** [comment]
I really have no clue how a FIFO works, but was hoping there was a way to interrupt prior previews if a new one is being issued, without waiting. Something I vaguely remember [@leovilok solved](https://github.com/jarun/nnn/issues/617#issuecomment-638518110) with a simple `&` when I was having a similar issue with `viu` (hence why I tried `&` here, but I guess it's only a partial solution and reaches its limits when preview load time is longer).

**Comment by luukvbaal** [comment]
I was actually able to reproduce the queuing to some extent with kitty on my laptop with key-repeat and #1006 should fix the issue.

If at all possible I would suggest using `ueberzug` instead of kitty's `icat` if speed is an issue for you since it is ALOT faster.

**Comment by jarun** [comment]
> If at all possible I would suggest using ueberzug instead of kitty's icat if speed is an issue for you since it is ALOT faster.

Completely agree! As someone who multitasks all the time my experience with kitty has been horrible. And I have started compiling with `O_FIFO=1` only after @luukvbaal added ueberzug support.

**Comment by KlzXS** [comment]
TIL github will match "fix #XXX" in a middle of a sentence.

**Comment by jarun** [comment]
Nice image BTW @KlzXS!

**Comment by jarun** [comment]
@luukvbaal and @Kabouik can you guys audit `nuke` + `preview-tui*` (all plugins in general) and see if it can be optimized?

Things I can think of:

- `case` in place of too many `if`s
- use of [shell builtins](https://tldp.org/LDP/abs/html/internal.html#BUILTINREF) (e.g. `type` instead of `which`)
- it would be great to get rid of `echo` for `printf`

This [optimization document](https://tldp.org/LDP/abs/html/optimizations.html) is a good start if you need any reference.

**Comment by luukvbaal** [comment]
Yeah some of those things have been bothering me since I started working on `preview-tui` but I didn't want to change the style you guys adopted. I'll see what can be done for `preview-tui` however I don't use `nuke` myself so I'd rather if someone else could pick that up.

**Comment by jarun** [comment]
> however I don't use nuke myself

Doesn't matter. Please have a look if you have the time.

**Comment by wustho** [comment]
@jarun hello there, sorry the tread is locked. About cleanfilename, with latest commit (merged) it won't overwrite any file...

Instead, it will be prepended with _ if file with the same name already exists... Eg. file1 --> _file1 if file1 already exists...

**Comment by jarun** [comment]
@wustho awesome! Thanks for the confirmation.

**Comment by Kabouik** [comment]
> > If at all possible I would suggest using ueberzug instead of kitty's icat if speed is an issue for you since it is ALOT faster.
> 
> Completely agree! As someone who multitasks all the time my experience with kitty has been horrible. And I have started compiling with `O_FIFO=1` only after @luukvbaal added ueberzug support.

Ha, thanks for the advice! That's frustrating after I spent time configuring kitty **exactly** because I wanted to use `icat` and finally have real image previews in terminal. I assumed `icat` being native to kitty, it would perform better than alternatives. I'll try the combination `kitty` + `ueberzug` on the slow machine instead.

@jarun what exactly was horrible for you with kitty? Overall performance regardless if `icat`?

**Comment by jarun** [comment]
> what exactly was horrible for you with kitty?

Image rendering.

**Comment by Kabouik** [comment]
I failed at prioritizing `ueberzug` over kitty's `icat` in the plugin (I shifted its close to the top but broke previews, won't bother you with that here), but to get a better idea of `ueberzug`'s performance, I compared image preview between a [`fff` fork using `ueberzug`](https://github.com/zeljkobekcic/fff) and [`fff` upstream using `w3mimgdisplay`](https://github.com/dylanaraps/fff).

I got much better rendering speed with `w3mimgdisplay`. Through my reading I learnt its advantages over `ueberzug` are it works on Wayland too, can print images on the tty, and doesn't need a daemon like `ueberzug` (apparently).

I am sure it has drawbacks too, `ueberzug` is vey popular. But would it make sense to add support for `w3mimgdisplay` in preview plugins? `nnn` has a broad audience that certainly includes slow systems (like SBCs), machines using Wayland, or others with just a tty, where both `ueberzug` and `kitty` would be unusable. Since `w3mimgdisplay` appears less restrictive and faster (at least I can distinguish it from `ueberzug` on a slow system, maybe not on a fast), I guess it's not too redundant. `viu` could work on all those machines but there's a gap in functionality.

**Comment by Kabouik** [comment]
`w3mimgdisplay` doesn't support `tmux` so problem solved!

**Comment by luukvbaal** [comment]
You're right, prioritzing `ueberzug` over `icat` in `kitty` requires more changes than shifting the order in `image_preview()`. See e.g. lines such as: `[ "$TERMINAL" != "kitty" ] && exists ueberzug && ueberzug_remove` in `preview_fifo()`.

`ueberzug` previews are pretty much instantaneous on the machines I use `nnn` on but I can believe `w3mimgdisplay` could be faster as its writting in C as opposed to python.

The problem I have with `w3mimgdisplay` is that I've never gotten it to work in any of the terminals I've used. This is also one of the main reasons `ueberzug` exists in the first place if you go by the [README](https://github.com/seebye/ueberzug#readme)(it's definitely not less restirictive than `w3mimgdisplay` atleast for `X11` environments.

Having said that, if you can manage to add support for `w3mimgdisplay` to `preview-tui` I say go ahead. Again, I've had nothing with problems with `w3mimgdisplay` and haven't touched it since `ueberzug` was created so I see no reason add it myself.

**Comment by luukvbaal** [comment]
@Kabouik This is the [diff](https://gist.github.com/luukvbaal/a2f30395287567539f076ce2c1478f2d) you would need for `ueberzug` previews in `kitty`.

**Comment by Kabouik** [comment]
Nice, thanks a lot @luukvbaal! It's not instantaneous on my slow device but I can feel an improvement indeed. The queuing is there though (which is expected, it was prevented just for `icat` as far as I understand). However I have mixed feelings about the queuing in this case: on the one hand it's ugly and can go bad if scrolling for several seconds continuously, on the other hand `ueberzug` appears to create some kind of cache of images it already previewed, which means this cache is fed by the queued files and later previews are almost instantaneous even on this machine.

**Comment by luukvbaal** [comment]
I do believe `ueberzug` does some kind of caching yeah but I would also suggest following 
`# Consider setting NNN_PREVIEWDIR to $XDG_CACHE_HOME/nnn/previews if you want to keep previews on disk between reboots` on a slow device if you havent already. If disk space is less of an issue that is.

This will place converted images in `$XDG_CACHE_HOME` instead of the default `$TMPDIR` which is lost after a reboot.

**Comment by Kabouik** [comment]
Will do. Disk space is an issue since this is a [phone](https://www.youtube.com/watch?v=8PnDZtKzsSc) but I know where to clear the cache when I reach the limit.

Probably out of scope since it's not the official plugin, but with the diff to prioritize `ueberzug` over `icat` in kitty, the preview does not properly follow layout. See here with patched `preview-tui-ext` (`ueberzug` previews) on the left, and vanillaa `preview-tui` (`icat` previews) on the right:

https://user-images.githubusercontent.com/7107523/118337752-bbd1e300-b514-11eb-9da4-dec49a9af01f.mp4

**Comment by luukvbaal** [comment]
Yeah `preview-tui` is not aware of the preview pane origin as is the case for `tmux`. I already accounted for this for horizontal splits in the initial diff but forgot about vertical splits.

Try the following `ueberzug_layer()`
```
ueberzug_layer() {
    if [ "$TERMINAL" = "kitty" ]; then
				if [ "$SPLIT" = "v" ]; then
						x=$(($(tput cols) / 2 + 1))
						y=0
				else
						x=0
						y=$(($(tput lines) / 2 + 1))
				fi
		else
				x=0
				y=0
		fi
    printf '{"action": "add", "identifier": "nnn_ueberzug", "x": %d, "y": %d, "width": "%d", "height": "%d", "scaler": "fit_contain", "path": "%s"}\n' "$x" "$y" "$1" "$2" "$3" > "$FIFO_UEBERZUG"
}
```

Updated the gist as well.

**Comment by Kabouik** [comment]
This didn't solve the issue from my first tests, but that was just a quick test while on the move, I'll test more thoroughly with more terminals and scenarios.

Unrelated: I mentionned in [977](https://github.com/jarun/nnn/pull/977#issuecomment-830308763) that scripts and vanilla functions outputs are garbled after several actions because they are displayed on top of each other. In `cmusq` I just added `\n` at the end of all prompts and this turned out to work decently for the use case. Would that be usable for `nnn` itself and possibly other scripts?

Example of garbled output with cpg here:
![Screenshot_20210516_002](https://user-images.githubusercontent.com/7107523/118413030-84427280-b69d-11eb-97c4-dd3245a6a65e.png)
(By the way, noob question, how do we copy/paste from `nnn` and accept overwrite for the whole selection?)

I tried adding `\n` after `overwrite?` at L584 of `nnn.c` but it didn't work, I suppose the issue might be with `cpg` instead.

It may seem only an aesthetic issue but sometimes, depending on the length of sentences, it can be unreadable and it may be impossible to tell if the output expects an answer, what are the accepted answers, unless experienced with daily `nnn` use already.

**Comment by jarun** [comment]
> Would that be usable for nnn itself and possibly other scripts?

You can do that for all the scripts. Just ensure you see the problem in the script and then you do the change.

>  how do we copy/paste from nnn and accept overwrite for the whole selection?

This is not an `nnn` issue. `nnn` uses the interactive option of `cp`. See `cp` options. If you have something you'll have to recompile the code with that.

> I tried adding \n after overwrite? at L584 of nnn.c but it didn't work

That prompt is from [advcpmv](https://github.com/jarun/advcpmv). You can try to add an `fflush()` after every progress bar ending and see if that works. Please raise a PR if it works for you.

**Comment by jarun** [comment]
@Kabouik the broken overwrit eprompt is fixed at https://github.com/jarun/advcpmv/commit/6adf5441b995c1bc614c51ffa0cb0e0545d23b94.

@mizlan unfortunately no such documentation exists. Please feel free to ask any questions you have. Also, I invited to the nnn devs group. Please join that. We generally discuss there.

**Comment by mizlan** [comment]
So sorry, but I am unaware of the dev group and can't seem to find it. What platform is it on? Gitter?

**Comment by Kabouik** [comment]
Awesome, thanks @jarun. I didn't even have the time to try what you suggested, can't keep up with your pace. :>

**Comment by jarun** [comment]
> but I am unaware of the dev group and can't seem to find it. 

Please check your mail. I sent you an invitation 1/2 days back.

**Comment by mizlan** [comment]
The invitation for was collaboration (push access) right? I do not see any sort of discussion group associated with it. Sorry for the inconvenience.

**Comment by jarun** [comment]
Not that one. I've resent the invitation just now. Please check.
